files/image_config/memory_statistics/memory_statistics_handler.py

```diff
@@ -0,0 +1,232 @@
+        """
+        Handle SIGHUP signal for reloading the configuration.
+        """
+        self.logger.log("Received SIGHUP, reloading configuration.", logging.INFO)
+        self.reloading.set()  # Trigger reload
+
+    def handle_sigterm(self, signum, frame):
+        """
+        Handle SIGTERM signal for graceful shutdown.
+        """
+        self.logger.log("Received SIGTERM, shutting down gracefully.", logging.INFO)
+        self.shutdown_event.set()  # Trigger shutdown
+
+    def format_memory_size(self, size):
+        """
+        Convert memory size to human-readable format (e.g., MB, GB).
+        """
+        for unit in ['B', 'KB', 'MB', 'GB', 'TB']:
+            if size < 1024:
+                return f"{size:.2f} {unit}"
+            size /= 1024
+
+    def collect_memory_statistics(self):
+        """
+        Collect memory statistics using psutil.
+        """
+        mem = psutil.virtual_memory()
+        return {
+            'total_memory': self.format_memory_size(mem.total),
+            'used_memory': self.format_memory_size(mem.used),
+            'free_memory': self.format_memory_size(mem.free),
+            'available_memory': self.format_memory_size(mem.available),
+            'cached_memory': self.format_memory_size(mem.cached),
+            'buffer_memory': self.format_memory_size(mem.buffers),
+            'shared_memory': self.format_memory_size(mem.shared)
+        }
+
+    def store_memory_statistics(self, memory_statistics):
+        """
+        Store memory statistics in a gzipped file.
+        """
+        try:
+            with gzip.open(self.filename, 'wt') as gz_file:
+                gz_file.write(f"{memory_statistics}\n")
+        except Exception as e:
+            self.logger.log(f"Error writing memory statistics to gzip file: {e}", logging.ERROR)
+
+    def cleanup_old_files(self):
+        """
+        Clean up old memory statistics log files.
+        Removes only .gz files, not the .log files.
+        """
+        for file in os.listdir(self.hdir):
+            if file.endswith('.gz'):
+                file_path = os.path.join(self.hdir, file)
+                if os.path.exists(file_path):
+                    try:
+                        os.remove(file_path)  # Remove the .gz log file
+                        self.logger.log(f"Deleted old log file: {file_path}", logging.INFO)
+                    except Exception as e:
+                        self.logger.log(f"Error deleting old log file: {e}", logging.ERROR)
+
+    def run_memory_collection(self):
+        """
+        Thread to collect and store memory statistics periodically.
+        """
+        while not self.shutdown_event.is_set():
+            try:
+                memory_statistics = self.collect_memory_statistics()
+                self.store_memory_statistics(memory_statistics)
+            except Exception as e:
+                self.logger.log(f"Error collecting or storing memory statistics: {e}", logging.ERROR)
```

Comment:
will this be repeated in dead loop and generate bunch of logs?

Reply:
Thank you for your feedback. The memory statistics daemon is designed with robust loop control to prevent dead loops. It collects data at a configurable interval, logs it in compressed format for historical data, and shuts down gracefully when triggered. The loop control mechanism ensures it waits according to the sampling interval and only continues running until a shutdown signal is received. Potential issues could arise from misconfigurations (e.g., zero or negative sampling intervals) but the daemon is designed to handle such cases effectively.